compositor_pipeline/src/audio_mixer/internal_audio_mixer.rs

```diff
@@ -189,4 +60,184 @@ impl InternalAudioMixer {
+            self.outputs
+                .iter()
+                .map(|(output_id, output_info)| {
+                    let samples = mix(&input_samples, output_info, samples_count);
+                    (output_id.clone(), OutputSamples { samples, start_pts })
+                })
+                .collect(),
+        )
+    }
+
+    fn merge_fill_input_samples(
+        &self,
+        input_samples_set: InputSamplesSet,
+    ) -> (HashMap<InputId, Vec<(i16, i16)>>, usize) {
+        let samples_count = (input_samples_set
+            .end_pts
+            .saturating_sub(input_samples_set.start_pts)
+            .as_secs_f64()
+            * self.output_sample_rate as f64)
+            .round() as usize;
+
+        let input_samples = input_samples_set
+            .samples
+            .into_iter()
+            .map(|(input_id, input_batch)| {
+                let samples = Self::frame_input_samples(
+                    input_samples_set.start_pts,
+                    input_samples_set.end_pts,
+                    input_batch,
+                    self.output_sample_rate,
+                );
+
+                (input_id, samples)
+            })
+            .collect();
+        (input_samples, samples_count)
+    }
+
+    fn frame_input_samples(
+        start_pts: Duration,
+        end_pts: Duration,
+        samples: Vec<InputSamples>,
+        sample_rate: u32,
+    ) -> Vec<(i16, i16)> {
+        let mut samples_in_frame = Vec::new();
+
+        samples
+            .into_iter()
+            .fold(start_pts, |last_end_pts, input_samples| {
+                let time_since_last_end = input_samples
+                    .start_pts
+                    .saturating_sub(last_end_pts)
+                    .as_secs_f64();
+                let missing_samples = (time_since_last_end * sample_rate as f64).round() as usize;
+                if missing_samples > 1 {
+                    Self::push_missing_samples(&mut samples_in_frame, missing_samples);
+                };
+                // Pushing InputSamples in frame
+                let sample_time_before_frame = start_pts
+                    .saturating_sub(input_samples.start_pts)
+                    .as_secs_f64();
+                let start_index = (sample_time_before_frame * sample_rate as f64).round() as usize;
+
+                let sample_time_after_frame =
+                    input_samples.end_pts.saturating_sub(end_pts).as_secs_f64();
+                let samples_after_frame =
+                    (sample_time_after_frame * sample_rate as f64).round() as usize;
+                let end_index = input_samples.len() - samples_after_frame;
+
+                samples_in_frame.extend(input_samples.samples[start_index..end_index].iter());
+
+                input_samples.end_pts
+            });
+
+        let missing_samples = Self::samples_in_frame(start_pts, end_pts, sample_rate);
+        Self::push_missing_samples(&mut samples_in_frame, missing_samples);
```

Comment:
I don't like how this works. It seems like a catch all case to hide any errors in the code above.

Reply:
No, it's here to fill samples in case the last batch for some input isn't pushed from the queue to the mixer. I'm talking about a case in which the input samples set looks like this:
```rust
InputSamplesSet {
    start_pts: 0ms,
    end_pts: 20ms,
    samples: {
           input_1: [Samples{start_pts: 0ms, end_pts: 10ms}]
    }
}
```
where the next batch for input_1 was received yet.

This should fill time [10ms to 20ms].

I also want to guarantee, that this function returns Vec with all samples needed for mixing in [set_start_pts, set_end_pts]. The previous loop can introduce rounding errors resulting in missing a few samples (calculating those indexes isn't (and I'm pretty sure it's extremely hard to make it) numerically safe).


Another approach would be to return the last end pts from this fold and fill the samples missing in [last_input_samples_end_pts, set_end_pts], but it doesn't deal with the index calculation errors mentioned above.

What do you exactly suggest here?